File: jiadaizhao_LeetCode/0501-0600/0503-Next Greater Element II/0503-Next Greater Element II.cpp
#include <bits/stdc++.h>
#include <vector>
#include <iostream>
#include <string>
#include <map>
#include <set>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) {
            return {};
        }
        stack<int> St;
        vector<int> result(n, -1);
        for (int i = 0; i < 2 * n - 1; ++i) {
            while (!St.empty() && nums[St.top()] < nums[i % n]) {
                result[St.top()] = nums[i % n];
                St.pop();
            }
                
            St.push(i % n);
        }
        
        return result;
    }
};

File: jiadaizhao_LeetCode/0501-0600/0503-Next Greater Element II/0503-Next Greater Element II.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        St = []
        n = len(nums)
        result = [-1] * n
        for i in range(2 * n - 1):
            while St and nums[St[-1]] < nums[i % n]:
                result[St.pop()] = nums[i % n]
            St.append(i % n)
        return result
